[10.0] DRAFT : [IMP] stock picking batch creation

lmarion-source opened the pull request.

lmignon committed:
[FIX] pre-commit: Add excluded files and fix missing website key into manifest

lmarion-source committed:
[IMP] stock picking batch creation

lmarion-source committed:
[IMP] stock picking batch creation

lmarion-source committed:
[IMP] stock picking batch creation

lmarion-source committed:
Merge branch '10.0-stock_picking_batch_creation' of github.com:acsone/wms into 10.0-stock_picking_batch_creation

lmarion-source committed:
fix typos

lmarion-source committed:
Put the sequence at the device type level -- remove device type line concept

lmarion-source committed:
[FIX] Check the device to be used at each picking (avoid clustering pickings that should no be on the same device)

lmarion-source committed:
[IMP] Add info about volume, weight and needed number of bins on the pickings and on the picking wave

lmarion-source committed:
[FIX] Take the user_id into account properly : start by looking for pickings for the user. If not, restart the search without user

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -0,0 +1,312 @@
+        )
+        if selected_pickings:
+            return selected_pickings, unselected_pickings
+        # 2) Check weither the first picking is breaking one condition or not
```
@lmarion-source Why do you stop on first picking it this one is not suitable for the batch may be the next one could be use... We should iterate over the pickings to find the first one and only if we don't find a picking launch an expeption...

lmignon committed:
[FIX] stock_picking_batch_creation: pre-commit

lmignon committed:
[FIX] tock_picking_batch_creation: Add dev requirements

lmignon committed:
[FIX] tock_picking_batch_creation: Add dev requirements

lmignon committed:
[IMP] stock_picking_batch_creation: Specific errror class + add possibility to silently return Nothing if a it's not possible to create a batch

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -0,0 +1,349 @@
+                break
+        if not device:
+            if raise_if_not_possible:
+                raise NoSuitableDeviceError()
```
@lmarion-source To improve readability, I would like to put all this logic into a specific_method:

```python
def _get_candidates_for_first_suitable_device(self, pickings):
    """Get the first suitable device and the list of pickings where all the
    pickings before the first one providing a suitable device are removed
    """
    excluded_ids = set()
    for pick in pickings:
        device = self._compute_device_to_use(pick)
        if device:
            return device, self.env["stock.picking"].browse([p.id for p in picking.ids if p.id not in excluded_ids])
       else:
           excluded_ids.add(pick.id)
    return self.env["stock.device.type"].browse(), self.env["stock.picking"].browse()
```

In the method:
``` python
device, picking = self._get_first_suitable_device_and_pick(candidates_pickings_to_batch)
if not device:
    ...

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -0,0 +1,349 @@
+            candidates_pickings = self._search_pickings()
+        return candidates_pickings
+
+    def _compute_device_to_use(self, first_picking_to_batch):
```
@lmarion-source The parameter should be `picking`here

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -0,0 +1,349 @@
+            weight_limit_outreached = device.max_weight and gt(
+                total_weight + picking.total_weight_batch_picking,
+                device.max_weight,
+                precision_weight,
+            )
+            volume_limit_outreached = device.max_volume and gt(
+                total_volume + picking.total_volume_batch_picking,
+                device.max_volume,
+                precision_volume,
+            )
+            nbr_lines_limit_outreached = (
+                self.maximum_number_of_preparation_lines
+                and gt(
+                    total_nbr_picking_lines + picking.nbr_picking_lines,
+                    self.maximum_number_of_preparation_lines,
+                    1,
+                )
+            )
+            available_bins_outreached = (
+                available_nbr_bins - picking.nbr_bins_batch_picking < 0
+            )
+
+            if not (
+                weight_limit_outreached
+                or volume_limit_outreached
+                or nbr_lines_limit_outreached
+                or available_bins_outreached
+                or different_device_for_current_picking
+            ):
+                # All conditions are OK : we keep the picking in the cluster
+                selected_pickings |= picking
+                total_weight += picking.total_weight_batch_picking
+                total_volume += picking.total_volume_batch_picking
+                total_nbr_picking_lines += picking.nbr_picking_lines
+                available_nbr_bins -= picking.nbr_bins_batch_picking
+            else:
+                # At least one condition KO : we discard the picking from the cluster
+                unselected_pickings |= picking
+
+        return selected_pickings, unselected_pickings
```
We should not have a specific code for the first picking and you should exclude a picking int the loop as of a condition is not met.

```suggestion
        selected_picking_ids = []
        total_weight = 0.0
        total_volume = 0.0
        total_nbr_picking_lines = 0
        precision_weight = self._precision_weight()
        precision_volume = self._precision_volume()

        def gt(value1, value2, digits):
            """Return True if value1 is greater than value2"""
            return tools.float_compare(value1, value2, precision_digits=digits) == 1

        # Always take the first picking even if it doesn't fullfill the
        # weight/volume criteria. This allows to always process at least
        # one picking which won't be processed otherwise
        available_nbr_bins = device.nbr_bins
        
        # Add pickings to the batch as long as
        # - total volume is not outreached
        # - total weight is not outreached
        # - maximum number of preparation lines is not outreached
        # - numbers of bins available is greater than 0
        # - The device for the current picking is supposed to be
        for picking in pickings:
            available_nbr_bins == 0:
                break
              
            if max_pickings and len(selected_pickings) == max_pickings:
                # selected enough!
                break
            
            # Check that the picking can go on the selected device
            picking_device = self._compute_device_to_use(picking)
            if device != picking_device:
                continue

             available_bins_outreached = (
                available_nbr_bins - picking.nbr_bins_batch_picking < 0
            )
            if available_bins_outreached :
                 continue

            picking.total_weight_batch_picking = self._picking_weight(picking)
            picking.total_volume_batch_picking = self._picking_volume(picking)
            picking.nbr_picking_lines = self._picking_nbr_lines(picking)
            picking.nbr_bins_batch_picking = math.ceil(
                picking.total_volume_batch_picking / volume_per_bin
            )

            weight_limit_outreached = device.max_weight and gt(
                total_weight + picking.total_weight_batch_picking,
                device.max_weight,
                precision_weight,
            )
            if weight_limit_outreached:
                continue 
            
            volume_limit_outreached = device.max_volume and gt(
                total_volume + picking.total_volume_batch_picking,
                device.max_volume,
                precision_volume,
            )
            if volume_limit_outreached:
                continue
            
            nbr_lines_limit_outreached = (
                self.maximum_number_of_preparation_lines
                and gt(
                    total_nbr_picking_lines + picking.nbr_picking_lines,
                    self.maximum_number_of_preparation_lines,
                    1,
                )
            )
            if nbr_lines_limit_outreached:
                continue
            
            # All conditions are OK : we keep the picking in the cluster
            selected_picking_ids.append(picking.id)
            total_weight += picking.total_weight_batch_picking
            total_volume += picking.total_volume_batch_picking
            total_nbr_picking_lines += picking.nbr_picking_lines
            available_nbr_bins -= picking.nbr_bins_batch_picking

        selected_pickings = self.env["stock.picking"].browse(selected_picking_ids)
        unselected_pickings = pickings - selected_pickings
        return selected_pickings, unselected_pickings
```

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -0,0 +1,349 @@
+            picking = first_picking
+        return picking
+
+    def _check_first_picking(self, first_picking, device):
```
Useless, should be removed

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -0,0 +1,349 @@
+        self._assign_operator_on_selected_pickings(batch, user)
+        return batch
+
+    def _check_number_of_available_bins(self, first_picking, device):
```
Useless, should be removed

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -0,0 +1,349 @@
+                raise NoSuitableDeviceError()
+            return self.env["stock.picking.wave"].browse()
+
+        candidates_pickings_to_batch = self._exclude_first_unfit_pickings_on_device(
```
This method code and the called method are now useless (see previous comment)

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -0,0 +1,349 @@
+    def _picking_nbr_lines(self, picking):
+        return len(picking.pack_operation_ids)
+
+    def _exclude_first_unfit_pickings_on_device(
```
To be remove since useless

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -0,0 +1,288 @@
+    def _picking_nbr_lines(self, picking):
+        return len(picking.pack_operation_ids)
+
+    def _get_first_suitable_device_and_pick(self, pickings):
```
```suggestion
    def _get_first_suitable_device_and_picks(self, pickings):
```

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -0,0 +1,288 @@
+            return self.env["stock.picking.wave"].browse()
+
+        device, picking = self._get_first_suitable_device_and_pick(
+            candidates_pickings_to_batch
```
```suggestion
        device, candidates_pickings_to_batch = self._get_first_suitable_device_and_picks(
            candidates_pickings_to_batch
```

lmignon reviewed (commented):
LGTM, 2 comments remains..

lmarion-source committed:
[FIX] Fix tests + Clean up code and refactore

lmignon committed:
[FIX] stock_picking_batch_creation: product_id is not required on pack operation.

lmignon committed:
[IMP] stock_picking_batch_creation: Avoid to pass method args by odoo context
This change is required to avoid perforamnces issue in 'product_qty_by_packaging' method

lmarion-source committed:
[FIX] Make sure that if a picking only contains products without dimensions, it will at least occupy one bin on the device

lmignon committed:
[IMP] stock_picking_batch_creation: Code cleanup + select first candidate even if one limit is exceed

lmarion-source committed:
[FIX] Take into account volume null for products in the computation of the cluster

lmarion-source committed:
[FIX] Correction on the nbr of bins occupied by pickings

lmarion-source committed:
[FIX] use related field for user_id on stock_picking

lmarion-source committed:
[IMP] Performances

lmarion-source committed:
[FIX] Only keep move lines that are available in the cluster for weight and volume computation

lmignon committed:
[ADD] stock_picking_batch_creation: Add method hook called once a picking is qualified to be added to the batch

IT allows you to lock the selected picking and ensure that nothing has changed in the timelapse between the search request and the addition to the batch that will prevent a normal commit of the transaction. If noting is returned by the method, the selected picking will not be added to the batch.

lmignon committed:
[IMP] stock_picking_batch_creation: Lock before update of stock.picking to avoid concurent update

lmignon commented on stock_picking_batch_creation/__manifest__.py:
```diff
@@ -14,6 +14,7 @@
     "application": False,
     "installable": True,
     "depends": [
+        "base_partition",
```
@lmarion-source This addon is not available on OCA....

lmignon replied:
Moreover it's not used at all...

lmarion-source replied:
My bad, I forgot to remove it!

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -156,6 +167,11 @@ def _create_batch(self, raise_if_not_possible=False):
         vals = self._create_batch_values(user, device, selected_pickings)
         batch = self.env["stock.picking.wave"].create(vals)
         batch._init_wave_info()
```
Why do you recompute the `wave_nbr_bins` information here and not pass the value to the `_init_wave_info(wave_nbr_bins)`

lmignon replied:
@lmarion-source The number of bins should be passed to the `init_wave_info` method

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -144,9 +153,11 @@ def _create_batch(self, raise_if_not_possible=False):
+        (
+            selected_pickings,
+            unselected_pickings,
+            volume_and_bins_by_partners,
```
we don't need to return volume and bins, we just need the number of used bins

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -191,6 +207,7 @@ def gt(value1, value2, digits):
         # # - numbers of bins available is greater than 0
         # # - The device for the current picking is supposed to be
         available_nbr_bins = device.nbr_bins
+        volume_and_bins_by_partners = defaultdict(list)
```
why `volume_and_bins` ????
```python

volume_by_partners = defaultdict(lambda: 0)
```

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -205,11 +222,21 @@ def gt(value1, value2, digits):
-            available_bins_outreached = (
-                available_nbr_bins - picking.nbr_bins_batch_picking < 0
-            )
+            if self.group_pickings_by_partner:
+                (
+                    bins_for_pickings_by_partner,
+                    volume_and_bins_by_partners,
+                ) = self._get_nbr_bins_by_partner(
+                    picking, device, volume_and_bins_by_partners
+                )
+                available_bins_outreached = (
+                    available_nbr_bins - bins_for_pickings_by_partner < 0
+                )
+            else:
+                available_bins_outreached = (
+                    available_nbr_bins - picking.nbr_bins_batch_picking < 0
+                )
```
seems too complicated and could be simplified...
```python 
    nbr_bins = picking.nbr_bins_batch_picking
    old_volume = volume_by_partner[picking.partner_id]
    new_volume = volume_by_partner[picking.partner_id] + picking.total_volume_batch_picking
    if self.group_pickings_by_partner:
        nbr_bins = math.ceil(new_volume / device.volume_per_bin) - math.ceil(old_volume / device.volume_per_bin)
    
    available_bins_outreached = (
        available_nbr_bins - nbr_bins < 0
    )
    if available_bins_outreached:
        continue
    ...
    ...
    available_nbr_bins -= nbr_bins
    volume_by_partner[picking.partner_id]  = new_volume

...
used_nbr_bins = device.nbr_bins - available_nbr_bins
return selected_pickings, unselected_picking, used_nbr_bins
```

moreover the volume was summed into the total by partner even if the picking was excluded due to some limits reached... You should add a test for this case....

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -245,11 +272,37 @@ def gt(value1, value2, digits):
-        return selected_pickings, unselected_pickings
+        return selected_pickings, unselected_pickings, volume_and_bins_by_partners
+
+    def _get_nbr_bins_by_partner(self, picking, device, volume_and_bins_by_partners):
+        if picking.partner_id.id in volume_and_bins_by_partners.keys():
+            old_nb_bins = volume_and_bins_by_partners[picking.partner_id.id][1]
+            new_volume = (
+                volume_and_bins_by_partners[picking.partner_id.id][0]
+                + picking.total_volume_batch_picking
+            )
+            new_nb_bins = math.ceil(new_volume / device.volume_per_bin)
+            if new_nb_bins > old_nb_bins:
+                bins_for_pickings_by_partner = new_nb_bins - old_nb_bins
+            else:
+                bins_for_pickings_by_partner = 0
+            volume_and_bins_by_partners[picking.partner_id.id] = [
+                new_volume,
+                new_nb_bins,
+            ]
+        else:
+            bins_for_pickings_by_partner = picking.nbr_bins_batch_picking
+            volume_and_bins_by_partners[picking.partner_id.id] = [
+                picking.total_volume_batch_picking,
+                picking.nbr_bins_batch_picking,
+            ]
+        return bins_for_pickings_by_partner, volume_and_bins_by_partners
```
must be removed since it's useless

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -205,11 +218,20 @@ def gt(value1, value2, digits):
+            nbr_bins = picking.nbr_bins_batch_picking
+            old_volume = volume_by_partners[picking.partner_id]
+            new_volume = (
+                volume_by_partners[picking.partner_id]
```
s/ volume_by_partners[picking.partner_id]/old_volume

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -227,6 +249,10 @@ def gt(value1, value2, digits):
                 precision_volume,
             )
             if volume_limit_outreached:
+                new_volume = (
+                    volume_by_partners[picking.partner_id]
+                    - picking.total_volume_batch_picking
+                )
```
@lmarion-source These lines are useless... Why do you need these?

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -191,6 +202,8 @@ def gt(value1, value2, digits):
         # # - The device for the current picking is supposed to be
         available_nbr_bins = device.nbr_bins
+        volume_by_partners = defaultdict(lambda: 0)
+        # volume_and_bins_by_partners = defaultdict(list)
```
@lmarion-source this line must be removed

lmarion-source committed:
[IMP] Allow to group pickings by partners in one bin

rousseldenis commented:
@lmarion-source Could you rebase after https://github.com/OCA/wms/pull/522 merge ?

This will allow to get test working

lmarion-source committed:
FIX pre-commit

lmignon commented:
@lmarion-source You must adapt the manifest to declare our addon as Alpha
```
stock_picking_batch_creation (Beta) depends on product_total_weight_from_packaging (Alpha)
```

lmarion-source committed:
Set development status to Alpha

lmignon commented on stock_picking_batch_creation/wizards/make_picking_batch.py:
```diff
@@ -222,6 +222,8 @@ def gt(value1, value2, digits):
             new_volume = old_volume + picking.total_volume_batch_picking
             if self.group_pickings_by_partner:
+                if not new_volume:
+                    new_volume = device.max_volume
```
@lmarion-source `new_volume = device.volume_per_bin`

lmarion-source committed:
[FIX] Handle the case of a volume of zero on pickings
